File: backend/app/schemas/violation.py
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, Field

from backend.app.models.violation import ViolationStatus, ViolationType


class ViolationCreate(BaseModel):
    camera_id: str
    violation_type: ViolationType
    license_plate: str | None = None
    confidence: float = Field(ge=0.0, le=1.0)
    location: str | None = None


class ViolationUpdate(BaseModel):
    status: ViolationStatus | None = None
    license_plate: str | None = None


class ViolationResponse(BaseModel):
    model_config = {"from_attributes": True}

    id: UUID
    camera_id: str
    violation_type: ViolationType
    status: ViolationStatus
    license_plate: str | None
    confidence: float
    fine_amount: float | None = 0.0
    clip_url: str | None
    thumbnail_url: str | None
    evidence_package_url: str | None
    location: str | None
    detected_at: datetime
    created_at: datetime


class ViolationStats(BaseModel):
    total_violations: int
    by_type: dict[str, int]
    by_status: dict[str, int]
    by_camera: dict[str, int]
    today_count: int
    this_week_count: int


class CameraCreate(BaseModel):
    id: str
    name: str
    stream_url: str
    location: str | None = None
    lat: float | None = None
    lng: float | None = None
    stop_line_geom: list[list[float]] | None = None


class CameraUpdate(BaseModel):
    name: str | None = None
    stream_url: str | None = None
    location: str | None = None
    status: str | None = None
    lat: float | None = None
    lng: float | None = None
    stop_line_geom: list[list[float]] | None = None


class CameraResponse(BaseModel):
    model_config = {"from_attributes": True}

    id: str
    name: str
    stream_url: str
    location: str | None
    status: str
    lat: float | None = None
    lng: float | None = None
    stop_line_geom: list[list[float]] | None = None
    last_frame_at: datetime | None = None
    created_at: datetime

File: frontend/src/types/violation.ts
export type ViolationType =
  | 'helmet_violation'
  | 'signal_jump'
  | 'wrong_way'
  | 'speeding'
  | 'no_seatbelt'
  | 'illegal_parking';

export type ViolationStatus =
  | 'detected'
  | 'confirmed'
  | 'rejected'
  | 'evidence_generated'
  | 'sent_to_authority';

export interface Violation {
  id: string;
  camera_id: string;
  violation_type: ViolationType;
  status: ViolationStatus;
  license_plate: string | null;
  confidence: number;
  fine_amount?: number;
  clip_url: string | null;
  thumbnail_url: string | null;
  evidence_package_url: string | null;
  location: string | null;
  detected_at: string;
  created_at: string;
}

export interface ViolationStats {
  total_violations: number;
  by_type: Record<string, number>;
  by_status: Record<string, number>;
  by_camera: Record<string, number>;
  today_count: number;
  this_week_count: number;
}

export interface Camera {
  id: string;
  name: string;
  stream_url: string;
  location: string | null;
  status: string;
  lat?: number | null;
  lng?: number | null;
  stop_line_geom?: number[][] | null;
  last_frame_at?: string | null;
  created_at: string;
}

export interface CameraHealth {
  id: string;
  name: string;
  lat: number | null;
  lng: number | null;
  status: string;
  last_frame_at: string | null;
}

export interface PeakHourPoint {
  hour: number;
  count: number;
}

export interface RevenueStats {
  days: number;
  total_inr: number;
}

export interface SampleClip {
  filename: string;
  label: string;
  description: string;
  expected_violations: string[];
  source?: string | null;
  source_url?: string | null;
  duration_seconds?: number | null;
}
